### ISSUE TITLE
[R-package] workflow for encoding unseen (new) data

### ISSUE THREAD

**Issue by gofford** [microsoft/LightGBM#3404]
Hi! 

For a model that is trained weekly and used to make daily predictions, what is the 'correct' method for applying a R-trained LightGBM model to new, unseen daily data? 

An alternative way to ask the same question is: do the same encoding rules need to be applied to the unseen new data, and if so, how should this be done? I can't find any examples of this.

I'm seeing an issue where a model performs very well on training and test data, but the response is essentially random when that model is saved, reloaded, and applied to new data. If that same unseen data is included in the test data the output is no longer random. Suspect it is something to do with the categorical encoding, but I want to rule out user error before pursuing.

Current workflow:

* train and evaluate the `model` as normal. This is trained weekly. 
* save the best iteration of the model to a text file using ` model$save_model_to_string(NULL)`.
* save the encoding rules used in training (a list) to an `rds` file.
* archive both model string (.txt) and the encoding rules (.rds) to S3.
* For new data (daily), load the model (text file) and encoding rules (`rds`) from S3. 
* Prepared new data according to the saved encoding generated from initial train. 
* Make predictions using the saved model.

This works from a code perspective. There are no errors. But as I said the response is nonsense. So,

1) is this a valid workflow?
2) If not, how should this be done?
3) Even if it is valid, is there a better way of doing this?

If the workflow is valid but the issue persists then I will investigate further and get an example together. Thanks!

**Comment by jameslamb** [comment]
hi @gofford , thanks for using `LightGBM`! Can you please clarify a few things for me?

1. What version of LightGBM are you using?
2. Can you show me in code what you mean by "the encoding rules"?
3. Are you using the parameter `categorical_features` in your call to `lgb.train()`?

It would help if you can share a code sample that describes what you mean by "encoding generated from initial train".

**Comment by gofford** [comment]
Hey @jameslamb,

1) I'm using LightGBM 3.0.0.

2) Apologies for the ambiguity on the explanation! 'encoding generated from the initial train' is the `rules` output of `lightgbm::lgb.convert_with_rules` when applied to the training set. An abridged version of my processing code is:

3) Yes, I am using the `categorical_features` parameter in the call to `lgb.train()`

Here's some code. I don't have a sharable dataset for this issue yet, so I'm afraid it's not replicable. Assuming that I'm not using the package incorrectly I will work on pulling together a replicable example. 

   ```
   # Assume I've got a data.table called 'data', and that 'target' denotes my response variable
   features <- setdiff(names(data), target) 
   cat_features <- intersect(features, names(data)[sapply(data, class) %in% c("factor", "character")])

   # recency based split; last 3 months are test 
   split_date <- lubridate::today() %m-% months(3)
   train_ids <- unique(data[firstsaledate < split_date]$id)

   # split to train/test
   train_index <- which(data$id %in% train_ids)
   train <- data[ train_index, c(target, features), with = FALSE]
   test <- data[-train_index, c(target, features), with = FALSE]

   # convert categoricals
   encoded_data <- lightgbm::lgb.convert_with_rules(data = train)
   train <- encoded_data$data
   test <- lightgbm::lgb.convert_with_rules(data = test, rules = encoded_data$rules)$data

   # split to feature matrices and target vectors
   train_X <- train[, features, with = FALSE]
   train_X <- as.matrix(train_X)
   train_y <- train[[target]]

   test_X <- test[, features, with = FALSE]
   test_X <- as.matrix(test_X)
   test_y <- test[[target]]

   # construct datasets
   dtrain <- lightgbm::lgb.Dataset(
       data = train_X,
       label = train_y,
       categorical_feature = cat_features,
       free_raw_data = FALSE
   )

   dtest <- lightgbm::lgb.Dataset.create.valid(
      dataset = dtrain,
      data = test_X,
      label = test_y
   )

   # train model
   model <- lightgbm::lgb.train(
      params = list(
         objective = 'binary', 
         metric = 'auc'
      ),
      data = dtrain,
      valids = list(
         dtrain = dtrain,
         dtest = dtest
      ),
      early_stopping_rounds = 10,
      eval_freq = 20,
      verbose = 1
   )
```

This is all pretty standard, works absolutely fine within the same R session, and performance is great. However I need to apply this model to new data on an ongoing basis so I need to archive the model. To do this, I save the `model` string and `encoded_data$rules` objects to S3. I'm using the model string as the `.rds` file saved by LightGBM is enormous, and I need things to work relatively fast.

```
   # save model string
   model_string <- model$save_model_to_string(NULL)
   readr::write_file(model_string, "model_string.txt")
   system("aws s3 cp model_string.txt s3://my/bucket/model_string.txt")

   # save list of encodings from training set so that it can consistently be applied to new data
   encoding_rules <- encoded_data$rules
   readr::write_rds(encoding_rules, "encoding_rules.rds")
   system("aws s3 cp encoding_rules.rds s3://my/bucket/encoding_rules.rds")
```

For new data, the process I'm using is to then load `model_string.txt` and `encoding_rules.rds` from S3, apply them to the new data with `lightgbm::lgb.convert_with_rules`, and make a prediction as normal.

```
   # retrieve from S3
   system("aws s3 cp s3://my/bucket/model_string.txt model_string.txt")
   system("aws s3 cp s3://my/bucket/encoding_rules.rds encoding_rules.rds")

   # load to R
   model <- lightgbm::lgb.load(model_str = "model_string.txt")
   encoding_rules <- readr::read_rds("encoding_rules.rds")

   # load new data, and apply encoding rules
   newdata <- data.table::fread("newdata.csv")
   encoded_newdata <- llightgbm::lgb.convert_with_rules(data = newdata, rules = encoding_rules)$data

   # make predictions
   newdata[, rebuyprob := predict(model, data = as.matrix(encoded_newdata))]
```

At this point, inspection shows that the predictions on the new data follow a completely different distribution to what was observed in the original training data which is very strange and contrary to expectation. If I retrain and include the "newdata" in the test set (within the same session), then the results are back to what are expected.

So my questions are now:

* Based on this code, is this procedure correct? and,
* Do predictions on new data need to be converted using the sale rules as the original training data (I have assumed yes)?

If I'm not doing anything obviously wrong then I will need to dig deeper into what is going on here.

Thanks for the help!

**Comment by gofford** [comment]
Rolling back to v2.3.1 resolved whatever was happening here, but why it was happening in v3.0.0 remains unclear.

**Comment by jameslamb** [comment]
Ok sorry for the delayed response!

> Based on this code, is this procedure correct?

@gofford this procedure all looks good to me!

> Do predictions on new data need to be converted using the same rules as the original training data (I have assumed yes)?

Yes, before predicting on new data, you should call `lgb.convert_with_rules()` using the same rules created during training. Since your first call to  `lgb.convert_with_rules()` didn't provide your own rules, any character or factor columns will have been assigned numeric values based on the order they were encountered in that training data.

<hr>

Can you please elaborate on "the predictions on the new data follow a completely different distribution to what was observed in the original training data"?

<hr>

Based on what you've provided so far, I have a few theories on what is going wrong and ideas on how to test them

**Problem 1: The `.save()` and `.load()` round-trip is broken**

Since you opened this issue, we've merged a PR that fixes several bugs in `.save()` for models: #3368. You could try cloning the most recent `master`, re-installing the R package, and trying your workflow again.

You could also try saving the model to local disk, then loading it again with `lgb.load()`. Call `predict()` on your test set from training, and compare that to the predictions on the test set from your original training session. If the results are identical, it probably means you have a data problem and not a `{lightgbm}` problem.

**Problem 2: Your new data is very different from your training data**

Use `summary()` on your training data and compare that to `summary()` on the new data to be scored. Are there any notable differences? That could mean that you have an issue with inconsistencies between your training data and new data. For example, maybe your new data has categorical features which have a lot of values that weren't seen during training, or maybe a numeric feature is in the wrong units of measure.


I hope that helps! I
